### PR TITLE
Add ConvNet tutorial

### PR DESCRIPTION
**Pull request by lilianabs**

### CODE REVIEW

**Comment by DhairyaLGandhi** [comment]
Can we get rid of the Args and it's handling?

**Comment by lilianabs** [comment]
Do you mean taking out that part?

**Comment by logankilpatrick** [comment]
@lilianabs @DhairyaLGandhi it would be great to get this merged and live, we can always improve later on

**Comment by lilianabs** [comment]
Sure! I agree that we can merge it and then continue improving :)

File: tutorials/_posts/2021-02-07-convnet.md [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+using Flux, Flux.Data.MNIST, Statistics
+using Flux: onehotbatch, onecold, logitcrossentropy
+using Base.Iterators: partition
+using Printf, BSON
+using Parameters: @with_kw
+using CUDA
+if has_cuda()
+   @info "CUDA is on"
+   CUDA.allowscalar(false)
+end
```
Comment (DhairyaLGandhi) [inline tutorials/_posts/2021-02-07-convnet.md:24]:
```suggestion
CUDA.allowscalar(false)
```

File: tutorials/_posts/2021-02-07-convnet.md [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+`make_minibatch` takes the following steps:
+
+* Creates the `X_batch` array of size `28x28x1x128` to store the mini-batches. 
+* Stores the mini-batches in `X_batch`.
+* One hot encodes the labels of the images.
+* Stores the labels in `Y_batch`.
+
+
+ 
+ `get_processed_data` loads the train and test data from Flux.Data.MNIST. First, it loads the images and labels of the train data set, and creates an array that contains the indices of the train images that correspond to each mini-batch (of size `args.batch_size`). Then, it calls the `make_minibatch` function to create all of the train mini-batches. Finally, it loads the test images and creates one mini-batch that contains them all.
```
Comment (DhairyaLGandhi) [inline tutorials/_posts/2021-02-07-convnet.md:63]:
```suggestion
 `get_processed_data` loads the train and test data from `Flux.Data.MNIST`. First, it loads the images and labels of the train data set, and creates an array that contains the indices of the train images that correspond to each mini-batch (of size `args.batch_size`). Then, it calls the `make_minibatch` function to create all of the train mini-batches. Finally, it loads the test images and creates one mini-batch that contains them all.
```

File: tutorials/_posts/2021-02-07-convnet.md [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+   test_set = make_minibatch(test_imgs, test_labels, 1:length(test_imgs))
+ 
+   return train_set, test_set
+ 
+end
+```
+<br>
+ 
+ 
+Now, we define the `build_model` function that creates a ConvNet model which is composed of *three* convolution layers (feature detection) and *one* classification layer. The input layer size is `28x28`. Each convolution layer reduces the size of the image by applying the Rectified Linear unit (ReLU) and MaxPool operations.
```
Comment (DhairyaLGandhi) [inline tutorials/_posts/2021-02-07-convnet.md:85]:
```suggestion
Now, we define the `build_model` function that creates a ConvNet model which is composed of *three* convolution layers (feature detection) and *one* classification layer. The input layer size is `28x28`. The images are grayscale, which means there is only *one* channel (compared to 3 for RGB) in every data point. Combined together, the convolutional layer structure would look like `Conv(kernel, input_channels => output_channels, ...)`. Each convolution layer reduces the size of the image by applying the Rectified Linear unit (ReLU) and MaxPool operations.
```

File: tutorials/_posts/2021-02-07-convnet.md [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+   ```julia
+    paramvec(m) = vcat(map(p->reshape(p, :), params(m))...)
+   ```
+<br>
+ 
+* `anynan` checks whether any element is NaN or not:
+ 
+   ```julia
+     anynan(x) = any(isnan.(x))
+   ```
```
Comment (DhairyaLGandhi) [inline tutorials/_posts/2021-02-07-convnet.md:137]:
```suggestion
 
* `anynan` checks whether any element of the params is NaN or not:
 
   ```julia
     anynan(x) = any(y -> any(isnan.(y)), x)
   ```
```

File: tutorials/_posts/2021-02-07-convnet.md [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+  
+   @info("Beginning training loop...")
+   best_acc = 0.0
+   last_improvement = 0
+   for epoch_idx in 1:args.epochs
+       # Train for a single epoch
+       Flux.train!(loss, params(model), train_set, opt)
+      
+       # Terminate on NaN
+       if anynan(paramvec(model))
```
Comment (DhairyaLGandhi) [inline tutorials/_posts/2021-02-07-convnet.md:190]:
```suggestion
       if anynan(Flux.params(model))
```

**Review by DhairyaLGandhi — COMMENTED** [review]
@lilianabs could you check if you are happy with the suggestions, and they work fine. If so, lets get this merged!

**Comment by lilianabs** [comment]
> @lilianabs could you check if you are happy with the suggestions, and they work fine. If so, lets get this merged!

Thank you for the suggestions @DhairyaLGandhi :) I've committed them all. Let's get it merged.

**Comment by DhairyaLGandhi** [comment]
Thanks!
